docs/specs/oidc.md

```diff
@@ -405,3 +406,83 @@ Silent Authentication
+
+OAuth protocol allows user to delegate access to clients, so that clients can
+act on behalf of the user.
+
+Usually, OAuth clients are third-party clients, with limited trust:
+- User need to give consent to a client before clients is allowed access.
+- Privileged user operations (e.g. change password) is usually not exposed
+  through APIs accessible through OAuth protocol.
+
+However, developers may want to give first-party clients full trust:
+- First-party clients do not need user consent in authorization process.
+- First-party clients have access to privileged user operations through a
+  special OAuth scope value (`https://authgear.com/scopes/full-access`).
+
+To designate an OAuth client as first-party client, developer may set
+`is_first_party` attribute to `true` in the corresponding client metadata.
+
+Developers should note the security implications for first-party clients:
+- Access tokens for first-party clients should not be passed to third-party
+  clients with limited trust.
+  (TODO: on-behalf-of flow https://tools.ietf.org/html/rfc7523)
+  (TODO: authenticate clients using client secret)
+  
+### Session Token
+
+For mobile first-party clients, developer may want to 'transfer' the
+user session from the native app (obtained through OAuth protocol) to
+web UI. In this case, developer may use access token to exchange for a
+one-time-use session token, which can be used to copy the user session
+to web user agent.
+
+First, the native app should perform authentication through standard
+OAuth flow to obtain an access token. However, refresh token must not
+be requested (i.e. not including `offline_access` in requested scopes),
+so that the access token would be bound to the user session instead of
+a refresh token.
+
+When the native app wants to copy the user session to web user agent,
+the native app may use the access token in the session token endpoint to
+obtain a session token:
+```
+POST /oauth2/session-token HTTP/1.1
+Host: accounts.example.com
+Authorization: Bearer <access token>
+
+---
+HTTP/1.1 200 OK
+Content-Type: application/json
+
+{"result":{"session_token":"<session token>"}}
+```
+
+A one-time-use session token would be returned, and the native app may
+then use it in OAuth authorization flow to copy user session to web user
+agent:
+```
+GET /oauth2/authorize?client_id=client_id&prompt=none&response_type=none
+    &login_hint=https%3A%2F%2Fauthgear.com%2Flogin_hint%3Ftype%3Dsession_token%26session_token%3D<session token>
+    &redirect_uri=<redirect URI> HTTP/1.1
+Host: accounts.example.com
+
+---
+HTTP/1.1 302 Found
+Set-Cookie: <session cookie>
+Location: <redirect URI>
+```
+
+When the session token is requested:
+- The OAuth client associated with the access token must be a first-party
+  client.
+- The access token must be bound to an IdP session (i.e. no refresh token)
+
+When the session token is consumed:
+- If the session token is invalid, normal OAuth authorization flow would be
```

Comment:
Any scenario which a client with valid access token might pass in an invalid session token? (So it should be a re-authentication?), or it is an error case?

Reply:
The session token is one-time-use and limited lifetime, so it can be already consumed, expired, or the associated session is invalidated/revoked/logout. Assume normal situation, these cases should not occur when consuming the token, since requesting the token would fail already.